jax/scipy/sparse/linalg.py

```diff
@@ -167,10 +183,254 @@ def cg(A, b, x0=None, *, tol=1e-5, atol=0.0, maxiter=None, M=None):
+  h = h.at[k + 1].set(v_norm)
+  H = H.at[k, :].set(h)
+  return V, H
+
+
+def apply_givens_rotations(H_row, givens, k):
+  """
+  Applies the Givens rotations stored in the vectors cs and sn to the vector
+  H_row. Then constructs and applies a new Givens rotation that eliminates
+  H_row's k'th element.
+  """
+  # This call successively applies each of the
+  # Givens rotations stored in givens[:, :k] to H_col.
+
+  def apply_ith_rotation(i, H_row):
+    cs, sn = givens[i, :]
+    H_i = cs * H_row[i] - sn * H_row[i + 1]
+    H_ip1 = sn * H_row[i] + cs * H_row[i + 1]
+    H_row = H_row.at[i].set(H_i)
+    H_row = H_row.at[i + 1].set(H_ip1)
+    return H_row
+
+  R_row = lax.fori_loop(0, k, apply_ith_rotation, H_row)
+
+  def givens_rotation(v1, v2):
+    t = jnp.sqrt(v1**2 + v2**2)
+    cs = v1 / t
+    sn = -v2 / t
+    return cs, sn
+  givens_factors = givens_rotation(R_row[k], R_row[k + 1])
+  givens = givens.at[k, :].set(givens_factors)
+  cs_k, sn_k = givens_factors
+
+  R_row = R_row.at[k].set(cs_k * R_row[k] - sn_k * R_row[k + 1])
+  R_row = R_row.at[k + 1].set(0.)
+  return R_row, givens
+
+
+def _gmres(A, b, x0, tol, n_kry, M):
+  """
+  Implements a single restart of GMRES. The n_kry-dimensional Krylov subspace
+  K(A, x0) = span(A(x0), A@x0, A@A@x0, ..., A^n_kry @ x0) is built, and the
+  projection of the true solution into this subspace is returned.
+  """
+  # https://www-users.cs.umn.edu/~saad/Calais/PREC.pdf
+  converged = False
+  residual = _sub(b, A(x0))
+  unit_residual, beta = _safe_normalize(residual, return_norm=True)
+  b_norm = jnp.sqrt(_vdot_tree(b, b, assume_real=False))
+
+  V = tree_map(
+    lambda x: jnp.pad(x[..., None], ((0, 0),) * x.ndim + ((0, n_kry),)),
+    unit_residual,
+  )
+  R = jnp.zeros((n_kry, n_kry + 1), jnp.result_type(*tree_leaves(b)))
+
+  givens = jnp.zeros((n_kry, 2), dtype=x0.dtype)
+  beta_vec = jnp.zeros((n_kry + 1), dtype=x0.dtype)
+  beta_vec = beta_vec.at[0].set(beta)
+
+  def loop_cond(carry):
+    k, err, _, _, _, _ = carry
+    return lax.cond(k < n_kry,
+                    lambda x: x[0]**2 > x[1],
+                    lambda x: False,
+                    (err, tol))
+    #return k < n_kry and err**2 > tol
+
+  def arnoldi_qr_step(carry):
+    k, err, V, R, beta_vec, givens = carry
+    V, H = kth_arnoldi_iteration(k, A, M, V, R, tol)
+    R_row, givens = apply_givens_rotations(H[k, :], givens, k)
+    R = R.at[k, :].set(R_row[:])
+    cs, sn = givens[k, :] * beta_vec[k]
+    beta_vec = beta_vec.at[k].set(cs)
+    beta_vec = beta_vec.at[k + 1].set(sn)
+    err = jnp.abs(sn) / b_norm
+    return k + 1, err, V, R, beta_vec, givens
+
+  carry = (0, beta, V, R, beta_vec, givens)
+  carry = lax.while_loop(loop_cond, arnoldi_qr_step, carry)
+  k, err, V, R, beta_vec, _ = carry
+  converged = k < n_kry - 1
+
+  y = jsp.linalg.solve_triangular(R[:, :-1].T, beta_vec[:-1])
+  dx = M(tree_map(lambda X: jnp.dot(X[..., :-1], y), V))
```

Comment:
This is a bit of a "gotcha" currently but if you're doing linear algebra on TPUs it's important to set `precision=lax.Precision.HIGHEST` whenever doing dot-product. Otherwise it defaults to bfloat16, which is usually not enough.

Reply:
Sounds good, thank you!